api/src/main/java/org/pmiops/workbench/workspaces/WorkspaceServiceImpl.java

```diff
@@ -130,22 +155,32 @@ public Workspace get(String ns, String firecloudName) {
         .collect(Collectors.toList());
   }
 
+  @Transactional
   @Override
-  public List<WorkspaceResponse> getWorkspaces() {
-    return getWorkspacesAndPublicWorkspaces().stream()
-        .filter(
-            workspaceResponse ->
-                !(workspaceResponse.getAccessLevel() != WorkspaceAccessLevel.OWNER
-                    && workspaceResponse.getAccessLevel() != WorkspaceAccessLevel.WRITER
-                    && workspaceResponse.getWorkspace().getPublished()))
-        .collect(Collectors.toList());
-  }
+  public WorkspaceResponse getWorkspace(String workspaceNamespace, String workspaceId) {
+    Workspace dbWorkspace = getRequired(workspaceNamespace, workspaceId);
 
-  @Override
-  public List<WorkspaceResponse> getPublishedWorkspaces() {
-    return getWorkspacesAndPublicWorkspaces().stream()
-        .filter(workspaceResponse -> workspaceResponse.getWorkspace().getPublished())
-        .collect(Collectors.toList());
+    org.pmiops.workbench.firecloud.model.WorkspaceResponse fcResponse;
+    org.pmiops.workbench.firecloud.model.Workspace fcWorkspace;
+
+    WorkspaceResponse response = new WorkspaceResponse();
+
+    // This enforces access controls.
+    fcResponse = fireCloudService.getWorkspace(workspaceNamespace, workspaceId);
+    fcWorkspace = fcResponse.getWorkspace();
+
+    if (fcResponse.getAccessLevel().equals(WorkspaceService.PROJECT_OWNER_ACCESS_LEVEL)) {
```

Comment:
I'd make a dedicated mapping function for this. A function like `getWorkspace` shouldn't need to know these details.

Reply:
Yeah... This is something I'm working on for the hackathon! Going to try to make a clear translation layer between our API, DB, and Firecloud models.

Reply:
Slick!